DHB Integration

HazelCC opened the pull request:
This PR integrate the data structure Dynamic Hashed Blocks (DHB) into Networkit. This improves NetworKit for dynamic graph operations.

### **TLTR:**

For a detailed description of DHB, refer to the paper: [*A Fast Data Structure for Dynamic Graphs Based on Hash-Indexed Adjacency Blocks* (2022)](https://drops.dagstuhl.de/storage/00lipics/lipics-vol233-sea2022/LIPIcs.SEA.2022.11/LIPIcs.SEA.2022.11.pdf)

**Why?**

Experiments have shown that DHB outperforms competing dynamic graph structures in edge insertions, updates, deletions, and traversal operations. This integration allows users to leverage DHB to gain performance for dynamic graph operations compared to the (vanilla) Graph class.

**How?**

This integration extends the functionalities of the default `Graph` class by supporting (almost) all methods implemented in `Graph`. When NetworKit is build using DHB, the necessary resources will be pulled in as a sub-module, and the associated code branches will be compiled. If the respective NetworKit modules are not compiled using the DHB integration, `DHBGraph` will fall back using the implementation of the default `Graph` class. 

**Measurement:**

Experiments run on commodity hardware (MacOS, M1, 8 GB RAM) have been run to get some baseline measurements on the performance of default `Graph` class vs. the `DHBGraph` class. For edge insertions, sequential weight update, and sequential edge removal, `DHBGraph` outperforms `Graph` class in an unweighted graph setting. 

![image(2)](https://github.com/user-attachments/assets/f9f81d0e-e8ac-4fed-a2a4-71db73ede511)
![image(1)](https://github.com/user-attachments/assets/5af9b4d2-d94d-4db7-a220-85f241a80104)
![image](https://github.com/user-attachments/assets/f49c42e0-abf8-428f-b4b9-5c133d4c63f5)

TODOs:

- [x]  ~~Review the passing by value semantics for all lambdas passed to a method such as `f(L handle)`. Instead, we might want to pass all lambdas by rvalue-reference as in `f(L&& handle)`.~~
- [x]  Review which type of `omp parallel for` scheduler (guided?) we use.
- [x]  ~~Review possibilities to use a concurrent/parallel version of functions that iterate over edges/vertices. For example, we might make `removeNode` to use the concurrent version of `forNeighbors`.~~
- [x]  Write more tests for `edgeIterator` and `nodeIterator`. (@fabratu)
- [x]  Write tests to check the consistency of ~~unweighted~~ undirected `DHBGraph` objects. An ~~unweighted~~ undirected graph should store one edge in both direction. (@HazelCC)
- [x]  Improve documentation. (@fabratu)
- [x] Fix GCC build issues. @c-bebop 
- [x] Build DHB on Clang and fix issues @c-bebop

Commit:
Integrating DHB represented by a new graph class.

Including all changes necessary for integrating the DHB graph data structure represented by a single graph class. Does not include tests nor necessary refactorings to reuse Edge type.

Commit:
Move Edge type to its own file.

This enables us to reuse the Edge type for the DHB Graph class.

Commit:
Test DHB Graph class functionality.

Commit:
Use east const!

fabratu commented on .clang-format:
```diff
@@ -88,7 +88,7 @@ PenaltyBreakFirstLessLess: 120
 PenaltyExcessCharacter: 1000000
 PenaltyReturnTypeOnItsOwnLine: 60
-PointerAlignment: Right
+PointerAlignment: Left
```
What is the rationale behind the clang-format changes?

c-bebop replied:
We can also revert those changes, if we want to make this another PR. The rationale is easily explained. To align the pointer symbol on right will clearly show that a type if of pointer or reference type as in:
```c++
int a = 10;
int* b = &a;
```  
Now `b` is of type _pointer to integer_ thus the symbol that makes it a pointer is part of the type not part of the variable name - as it would be with right alignment. The same applies to references. 

The `QualifierAlignment` in line `123` makes sure we are using _east const_ which simply follows the binding of a qualifier to it's left as in:
```c++
int const * const c = &a
```
Where the _const_ qualifier always binds to it's left. If one would want to use _west const_ one has to write:
```c++
const int * const c = &a
```
Basically mixing both alignments (west & east) since it is not possible to write:
```c++
const int const *  c = &a
```

fabratu commented on CMakeLists.txt:
```diff
@@ -18,6 +18,7 @@ option(NETWORKIT_CLANG_TIDY "Check code with clang-tidy" OFF)
 option(NETWORKIT_FLATINSTALL "Install into a flat directory structure (useful when building a Python package)" OFF)
 option(NETWORKIT_COVERAGE "Build with support for coverage" OFF)
 option(NETWORKIT_SANITY_CHECKS "Algorithms will perform further (expensive) checks" OFF)
+option(NETWORKIT_USE_DHB "Enable DHB)" ON)
```
`DHB` is set to enabled by default here. Does this also mean, that it is used instead of `Graph` as data structure by default?

c-bebop replied:
No it is not used by default. Instead there is a `DHBGraph` class now that can be instantiated. We did not change the default `Graph` class.

fabratu replied:
I see. When first reading into the PR I misunderstood how `DHB` is integrated.

fabratu commented on CMakeLists.txt:
```diff
@@ -265,6 +285,7 @@ if(NETWORKIT_BUILD_CORE AND NETWORKIT_MONOLITH)
 	endif()
 
 	target_link_libraries(networkit PRIVATE OpenMP::OpenMP_CXX PUBLIC tlx)
+	target_link_libraries(networkit PRIVATE dhb)
```
This again makes `dhb` mandatory. What is the point of using an option then?

c-bebop replied:
So the question here is rather: do we want to use `if` branches for example to remove the mandatory linkage to `dhb` or is there another way?

fabratu replied:
Ok, I revoke my wrapper assumption. Maybe I again misunderstood the concept of integrating the library. `DHBGraph.hpp` seems to include two cases: one implementation where `dhb` is available and one where it is not? 

If so, what is the functionality of `DHBGraph.hpp`/`DHBGraph.cpp` if the library is not present? Seems very similar to graph then, right?

fabratu reviewed (dismissed):
In order to keep the PR manageable, I suggest to fix all remaining issues concerning the integration in this PR and move optimization and additional thoughts to another PR.

Concerning building / testing: It seems that `DHB` is not correctly added as a submodule or there is some ill-defined logic in CMakeLists.txt. CI does fetch all submodules, however `DHB` is not found.

Commit:
set using dhb as false by default

Commit:
Add DHB as a submodule

c-bebop committed:
No option to pull in DHB anymore: pull in DHB is mandatory.

c-bebop committed:
Remove compile option for using DHB.

c-bebop committed:
Remove all code branches where DHB would not be used.

c-bebop committed:
Use omp parallel for schedule(guided) if possible.

c-bebop commented:
@HazelCC Could you clarify and elaborate what exactly you mean with the point: _Write tests to check the consistency of unweighted DHBGraph objects._?

c-bebop commented:
> > @HazelCC Could you clarify and elaborate what exactly you mean with the point: _Write tests to check the consistency of unweighted DHBGraph objects._?
> 
> An undirected graph object should store directed edges in both directions. Previously there were issues where logical bugs causing inconsistency for undirected graph, therefore it's good to have more tests addressing this.

For now, we have referenced you as taking over for this task. Is that fine?

c-bebop committed:
Address cmake policy CMP0077 by stating that DHB_WITH_64BIT_IDS is a cached boolean value.

c-bebop committed:
Pull in latest changes from DHB repo.

c-bebop committed:
addEdge() looses multi edge parameter (unused, not supported).

c-bebop committed:
Signed/unsigned compare warning as error fix.

c-bebop committed:
Remove unused variables.

c-bebop committed:
TODO: implement getIthNeighborWithWeight()

c-bebop committed:
Use DHB super_pedantic: latest commit.

c-bebop committed:
Revise doc: swapEdge().

c-bebop committed:
New private method addEdge() including parameter edge id.

c-bebop committed:
addEdge(u, v, ew) calls private addEdge() impl.

c-bebop committed:
Fix swapEdge().

c-bebop committed:
Move Edge header to graph directory.

c-bebop committed:
Revert "Use east const!"

This reverts commit 6c7f319fd5c81fa0e0f51e5b5003b9c1b4eadca8.

c-bebop committed:
Reapply clang format rules on dhb graph + tests.

c-bebop committed:
Use DHB system allocator for all MSVC targets when using DHB.

c-bebop committed:
Use latest commit including many GCC warnings as errors + mmap() approach for MSVC.

c-bebop committed:
Fix several issues with unused variables / parameters.

c-bebop committed:
Remove inEdges field.

c-bebop committed:
Remove inEdgeIds field.

c-bebop committed:
We do not need an edge merge operator.

c-bebop committed:
Fix correct field order for correct init order in DHBGraph() constructor.

c-bebop committed:
Do not put all DHBGraph tests in NetworKit namespace: instead, use namespace NetworKit within the test file.

c-bebop committed:
Fix swapEdges() test.

c-bebop committed:
Add blank line.

c-bebop committed:
Fix DHB testEdgeIterator test.

c-bebop committed:
Fix: add large graph test now uses ErdoRenyi generator to generate edges.

c-bebop committed:
Adding TODO for a static cast of dhb::EdgeID to NetworKit ID.

c-bebop committed:
Add generators module to DHBGraphGTest.

c-bebop committed:
Remove redundant include of DHBGraph.

c-bebop committed:
Change order of linking to dhb (and networkit module).

c-bebop committed:
Provide a comment on how a test target name is composed.

c-bebop committed:
Comment on how to add a NK module to a test target.

c-bebop committed:
Link against dhb publicly (for each module).

c-bebop committed:
Format code.

c-bebop committed:
Use DHB build fixing redundant semicolon.

c-bebop committed:
Revise forNodeParisParallel() implementation: fix issue of calling for wrong pair (u, v).

c-bebop committed:
Format.

c-bebop committed:
Actually no exception is thrown even if u > n.

c-bebop committed:
Use latest DHB commit fixing missing include of header numeric.

c-bebop committed:
Use omp index for omp parallel for.

c-bebop committed:
Implement getIthNeighbor() correctly.

c-bebop committed:
Return default edge weight if ith neighbor does not exist.

c-bebop committed:
Get and set of ith in neighbor is irrelevant for DHB.

c-bebop committed:
Use NK style include guard.

c-bebop committed:
Use omp_index for parallel region in forNodesParallel().

c-bebop committed:
Leave comment for #endif of include guard.

Commit:
Use signed index type variables for OpenMP parallel for-loops.

c-bebop committed:
Fix include guard code-style.

c-bebop committed:
Do not declare edge id as const in header.

c-bebop committed:
Use std::forward() instead of std::move().

c-bebop committed:
Cast num_threads value to int explicitly.

c-bebop committed:
Copy lambdas (not move them).

c-bebop committed:
Use push back for weighted edge.

c-bebop committed:
Small code review of inserting weighted edges.

c-bebop committed:
addEdges() will not return boolean indicating successful inserts.

In multi-threaded contexts, this introduces data dependencies to the variables a thread has to write the results to. We could make sure this is only written to one cache line but this seems overkill: the user can check for successful insertions post addEdges() call.

c-bebop committed:
Test for adding edges twice, once without update so there is no effect.

c-bebop committed:
addEdges() will not receive an arbitrary parameter for thread count.

Instead we rely on the systems (omp) assigned thread count of the machine.

c-bebop committed:
Fix type comparison warning in Debug mode.

c-bebop commented:
This PR (NetworKit DHB Integration) is now ready for review. All CI defined **builds are running successfully**. For further development please **address failing builds immediately**.

There are a couple of TODOs in code as well as in the PR (comments/description). @fabratu  and @HazelCC please make sure you will either address these as soon as possible, or leave comments in code for further development / improvements.

c-bebop commented on CMakeLists.txt:
```diff
@@ -390,7 +410,8 @@ function(networkit_add_module modname)
 			COMPILE_FLAGS "${NETWORKIT_CXX_FLAGS}"
 			LINK_FLAGS "${NETWORKIT_LINK_FLAGS}")
 
-		target_link_libraries(${MODULE_TARGET} PRIVATE OpenMP::OpenMP_CXX PUBLIC tlx)
```
For the `networkit` target we link `PRIVATE` for the module target we link `PUBLIC`. In `DHBGraph.hpp` we include DHB. So we must link publicly. We're not getting any errors here in our CI.. Perhaps, we never use `DHBGraph` in combination with a target linking against module `graph`?

fabratu replied:
`DHBGraph.hpp` is part of the `graph` module. Outside of the `graph` module we don't use the `dhb`-api directly, therefore it should not be necessary to link against `dhb` in the first place. Have you tried making `dhb` private for the module_target?

c-bebop commented on include/networkit/graph/DHBGraph.hpp:
```diff
@@ -0,0 +1,2882 @@
+ * @ingroup graph
+ * A graph (with optional weights) and parallel iterator methods.
+ *
+ * TODO:
```
TODO!

fabratu replied:
What is meant with "review" here? Is it about correctness? If not and this is about style/performance, I suggest remove any improvements (there are several TODOs in the code) to future PRs.

c-bebop replied:
Fine for me!

Commit:
Refactor edgeIterator tests (mimicing tests for Graph.hpp except for random edges.)

Commit:
Add nodeIterator tests (mimicing tests for Graph.hpp except for random nodes).

Commit:
Use generic Attributes template and refactor DHBGraph constructors.

Commit:
Remove datastructures / helper functions only needed for Graph-class.

Commit:
Remove more remnants of helpers, datastructures copied from Graph-class.

Commit:
Remove unneccessary and reorder remaining core member variables.

Commit:
Use common API for edge and node iterators.

Commit:
Remove NeighborIterators import

Commit:
Refactor forNodesParallel to be consistent with Graph API.

Commit:
Disable parallelism for removeSelfLoops test.

Commit:
Add missing reorder in constructor.

Commit:
Refactor edge count to be consistent with the Graph class.

Commit:
Refactor initialization by edge list and DHBGraph copy.

Commit:
Refactor addEdges to comply with standard NetworKit API.

Commit:
Remove outdated TODOs.

Commit:
Refactor edgeById.

Commit:
Refactor forNeighborsOf, forInNeighborsOf tests, increasing coverage.

Commit:
Refactor forInEdgesOf, increasing coverage.

Commit:
Refactor parallelSumForEdges, increasing coverage.

Commit:
Fix construction by initializer list + test forced index_edges, increasing coverage.

Commit:
Test non-existing neighbor with id.

Commit:
Test non-existing neighbor with weight.

Commit:
Test non-existing neighbor.

Commit:
Enable increaseWeight test.

Commit:
Add test for negative hasEdge result.

Commit:
Add corner case test for swapEdge.

Commit:
Test addEdges with update.

Commit:
Add test for weightedInDegree.

Commit:
Remove unneccessary check.

Commit:
Add test for calling edgeId on non-indexed graph.

Commit:
Test indexEdges twice for checking early return.

Commit:
Fix error in swapEdges, corner case test.

Commit:
Restrict addEdges test (weighted, no update) to be executed in a non-parallel way.

Commit:
Support updates when calling addEdge (+ adding tests).

c-bebop commented on CMakeLists.txt:
```diff
@@ -551,17 +575,23 @@ function(networkit_add_extra IS_TEST MOD NAME)
-					tlx
 					OpenMP::OpenMP_CXX
+					tlx
+					dhb
```
When linking publicly against `dhb` we may not need to link against `dhb` again here for the tests.

c-bebop commented on networkit/cpp/graph/test/DHBGraphGTest.cpp:
```diff
@@ -0,0 +1,2543 @@
+    });
+}
+
+TEST_P(DHBGraphGTest, testSetWeight) {
```
This test could also be separated into more tests .. as the comments suggest.

c-bebop reviewed (dismissed):
Looks very good, now just some small fixes.

fabratu commented on networkit/cpp/graph/DHBGraph.cpp:
```diff
@@ -0,0 +1,509 @@
+
+edgeid DHBGraph::edgeId(node u, node v) const {
+    if (!edgesIndexed) {
+        throw std::runtime_error(
```
I would rather stay with the current logic in order to mimic the behavior from `Graph`.

fabratu commented on networkit/cpp/graph/test/DHBGraphGTest.cpp:
```diff
@@ -0,0 +1,2543 @@
+    ASSERT_EQ(this->Ghouse.numberOfEdges(), DW.numberOfEdges());
+    this->Ghouse.forNodes([&](node v) {
+        count d = this->Ghouse.degree(v);
+        // count dUndirected = isDirected() ? d + this->Ghouse.degreeIn(v) : d;
```
Remnants of previous versions, deleted them.

fabratu commented on networkit/cpp/graph/test/DHBGraphGTest.cpp:
```diff
@@ -0,0 +1,2543 @@
+            bool hasEdge = containsEdge(edge);
+            bool hasEdgeReverse = containsEdge(edgeReverse);
+            if (this->Ghouse.isDirected()) {
+                ASSERT_EQ(hasEdge, this->Ghouse.hasEdge(u, v));
```
I agree, that using a tracking variable might reduce the number of asserts and make the test run faster. However, in the current state I don't see a running time problem. In addition I disagree with the argument concerning the improvement of error output. Currently the test is written in a way, that we know exactly for which edge the `assert` fails and therefore it makes it easier to search for bugs/problems. With a tracking variable, you would lose this information or would have to go through an array of booleans.

Commit:
Address code related review-comments.

bernlu commented on networkit/cpp/graph/DHBGraph.cpp:
```diff
@@ -0,0 +1,509 @@
+        --m;
+        return true;
+    }
+    return false;
+}
+
+void DHBGraph::removeAllEdges() {
+    m_dhb_graph = dhb::Matrix<EdgeData>(m_dhb_graph.vertices_count());
+    m = 0;
+    storedNumberOfSelfLoops = 0;
+    omega = 0;
+}
```
this function should invalidate all edgeAttributes as well.

c-bebop replied:
Done.

bernlu commented on include/networkit/graph/DHBGraph.hpp:
```diff
@@ -0,0 +1,1991 @@
+    /**
+     * Return edge weight of edge {@a u,@a v}. Returns 0 if edge does not
+     * exist.
+     * BEWARE: Running time is \Theta(1) for DHB implementation but \Theta(deg(u)) for fallback
+     * NetworKit graph DS.
```
i think this fallback nk graph is outdated information?

c-bebop replied:
True. Fixed.

bernlu commented on include/networkit/graph/DHBGraph.hpp:
```diff
@@ -0,0 +1,1991 @@
+     *
+     * @param u Node.
+     * @param i index; should be in [0, degreeOut(u))
+     * @return @a edge weight to the i-th (outgoing) neighbor of @a u, or @c +inf if no such
```
The function actually returns the default edge weight and not inf when the neighbor does not exist.

c-bebop replied:
I return `0.0` as in `nullWeight` now for all functions like `getIthNeighbor...()` which aligns to what you would expect from an edge that does not exist? :cowboy_hat_face:

bernlu commented on include/networkit/graph/DHBGraph.hpp:
```diff
@@ -0,0 +1,1991 @@
+     * @param u Node.
+     * @param i index; should be in [0, degreeOut(u))
+     * @return pair: i-th (outgoing) neighbor of @a u and the corresponding
+     * edge weight, or @c defaultEdgeWeight if unweighted.
```
we should add that `none` is returned when the neighbor does not exist.

c-bebop replied:
True. :pray: Done.

bernlu commented on include/networkit/graph/DHBGraph.hpp:
```diff
@@ -0,0 +1,1991 @@
+     *
+     * This function is not supported for DHB. Please use getIthNeighborWithWeight(node u, index i).
+     */
+    std::pair<node, edgeweight> getIthNeighborWithWeight(Unsafe, node u, index i) const;
```
why is this function declared twice, but once we claim that it is not supported while the other one is implemented?

c-bebop replied:
It's actually just this one which is not implemented. Marked this one as `deleted`.

bernlu reviewed (dismissed):
We should probably discuss some of these comments, some are quite detailed and maybe out of scope for this PR.

An additional general comment: I think it would be nice to add time complexity to the documentation for (almost) all functions.

c-bebop commented:
There are still some review comments but.. are we merging this one?

fabratu commented:
@c-bebop This is just a gentle reminder :)

invalid-email-address committed:
Remove old doc: semantics are derived from code.

c-bebop commented on include/networkit/graph/Edge.hpp:
```diff
@@ -0,0 +1,62 @@
+/**
+ * A weighted edge used for the graph constructor with
+ * initializer list syntax.
+ */
```
Agreed. I removed the doc string.

c-bebop commented on include/networkit/graph/Edge.hpp:
```diff
@@ -0,0 +1,62 @@
+struct WeightedEdge : Edge {
+    edgeweight weight;
+
+    // Needed by cython
```
What about `std::vector`?

c-bebop committed:
Remove forwarded CurveballMaterialization: unused.

c-bebop commented on include/networkit/graph/DHBGraph.hpp:
```diff
@@ -0,0 +1,1991 @@
+// forward declaration to randomization/CurveballImpl.hpp
+namespace CurveballDetails {
+class CurveballMaterialization;
+}
```
Seems so. Removed.

c-bebop committed:
Re-order branching and boolean logic for readability.

c-bebop commented on networkit/cpp/graph/DHBGraph.cpp:
```diff
@@ -0,0 +1,509 @@
+        is_inserted_in = m_dhb_graph.insert(v, u, edge_data, do_update);
+    } else {
+        is_inserted = m_dhb_graph.insert(u, v, edge_data, do_update);
+    }
```
Agree with that. Done.

c-bebop committed:
Moving DHB Graph init to init list.

c-bebop commented on networkit/cpp/graph/DHBGraph.cpp:
```diff
@@ -0,0 +1,509 @@
+      edgesIndexed(edgesIndexed), // edges are not indexed by default
+      nodeAttributeMap(this), edgeAttributeMap(this) {
+
+    m_dhb_graph = dhb::Matrix<EdgeData>(n);
```
:pray: Done.

c-bebop committed:
Revise linking target networkit to dhb, tlx as PUBLIC, OpenMP as PRIVATE.

c-bebop commented on CMakeLists.txt:
```diff
@@ -264,7 +280,11 @@ if(NETWORKIT_BUILD_CORE AND NETWORKIT_MONOLITH)
+									dhb 
+									OpenMP::OpenMP_CXX 
+									PUBLIC 
+									tlx)
```
Done.

c-bebop committed:
Revise DHB graph doc and add it's in experimental stage.

c-bebop commented on include/networkit/graph/DHBGraph.hpp:
```diff
@@ -0,0 +1,1991 @@
+ * TODO:
+ * - [ ] Review the passing by value semantics for all lambdas passed
+ *       to a method such as f(L handle). Instead, we might want to pass
+ *       all lambdas by rvalue-reference as in f(L&& handle).
```
Done. Revised doc of DHB graph class as well.

c-bebop committed:
delete Unsafe getIthNeighborWithWeight()

c-bebop committed:
Revise doc. of getIthNeighborWithWeight().

What if neighbor does not exists? Returns none.

c-bebop committed:
Re-assign Node/Edge Attribute.

c-bebop committed:
Add blank lines for read ability.

c-bebop committed:
Format update.

c-bebop committed:
Revise return values if edge does not exist.

c-bebop committed:
Revise doc on weight() fun.

No fallback solution to Graph.

c-bebop committed:
Fix typo in "missing module test target" comment.

c-bebop reviewed (dismissed):
The current state looks good to me. 🍏

Commit:
Update code style to clang format 22.

Commit:
Modify iterators to accept both templated and none templated graph types.

Commit:
Update code style to clang format 22.

Commit:
Fix errors in tests.

Commit:
Fix clang-tidy complaints.

fabratu commented:
Maybe there are still some sub-optimal parts in the code + the class is not templated. Since this PR has been open for a very long time, we will merge it now and deal with open ends later on.

c-bebop commented:
🥳 🎉

c-bebop commented:
Thank you @HazelCC @fabratu and @bernlu for your contributions to integrate DHB into Networkit.
